tests/azure_iot_hub_client_ut.c

```diff
@@ -381,6 +475,690 @@ static void testAzureIoTHubClient_SendTelemetry_Success( void ** state )
+                      eAzureIoTHubClientSuccess );
+}
+/*-----------------------------------------------------------*/
+
+static void testAzureIoTHubClient_CloudMessageSubscribe_InvalidArgFailure( void ** state )
+{
+    AzureIoTHubClient_t xTestIoTHubClient;
+
+    ( void ) state;
+
+    /* Fail SubscribeCloudToDevice when client is NULL */
+    assert_int_equal( AzureIoTHubClient_SubscribeCloudToDeviceMessage( NULL,
+                                                                       prvTestCloudMessage,
+                                                                       NULL, ( uint32_t ) -1 ),
+                      eAzureIoTHubClientInvalidArgument );
+
+    /* Fail SubscribeCloudToDevice when function callback is NULL  */
+    assert_int_equal( AzureIoTHubClient_SubscribeCloudToDeviceMessage( &xTestIoTHubClient,
+                                                                       NULL, NULL, ( uint32_t ) -1 ),
+                      eAzureIoTHubClientInvalidArgument );
+}
+/*-----------------------------------------------------------*/
+
+static void testAzureIoTHubClient_CloudMessageSubscribe_SendFailure( void ** state )
+{
+    AzureIoTHubClient_t xTestIoTHubClient;
+
+    ( void ) state;
+
+    prvSetupTestIoTHubClient( &xTestIoTHubClient );
+
+    will_return( AzureIoTMQTT_Subscribe, eAzureIoTMQTTSendFailed );
+    assert_int_equal( AzureIoTHubClient_SubscribeCloudToDeviceMessage( &xTestIoTHubClient,
+                                                                       prvTestCloudMessage,
+                                                                       NULL, ( uint32_t ) -1 ),
+                      eAzureIoTHubClientSubscribeFailed );
+}
+/*-----------------------------------------------------------*/
+
+static void testAzureIoTHubClient_CloudMessageSubscribe_ReceiveFailure( void ** state )
+{
+    AzureIoTHubClient_t xTestIoTHubClient;
+
+    ( void ) state;
+
+    prvSetupTestIoTHubClient( &xTestIoTHubClient );
+
+    will_return( AzureIoTMQTT_Subscribe, eAzureIoTMQTTSuccess );
+    will_return( AzureIoTMQTT_ProcessLoop, eAzureIoTMQTTRecvFailed );
+    assert_int_equal( AzureIoTHubClient_SubscribeCloudToDeviceMessage( &xTestIoTHubClient,
+                                                                       prvTestCloudMessage,
+                                                                       NULL, ( uint32_t ) -1 ),
+                      eAzureIoTHubClientFailed );
+}
+/*-----------------------------------------------------------*/
+
+static void testAzureIoTHubClient_CloudMessageSubscribe_Success( void ** state )
+{
+    AzureIoTHubClient_t xTestIoTHubClient;
+
+    ( void ) state;
+
+    prvSetupTestIoTHubClient( &xTestIoTHubClient );
+
+    will_return( AzureIoTMQTT_Subscribe, eAzureIoTMQTTSuccess );
+    will_return( AzureIoTMQTT_ProcessLoop, eAzureIoTMQTTSuccess );
+    xPacketInfo.ucType = azureiotmqttPACKET_TYPE_SUBACK;
+    xDeserializedInfo.usPacketIdentifier = usTestPacketId;
```

Comment:
Not directly related to this line, but code in question isn't in PR.  For the cmocka mock `AzureIoTMQTT_Init` and others, there's a global called `xTestUserCallback`.  This really confused me when trying to follow the path, because I assumed it meant user as far as getting to application layer and I couldn't see how it fit into the mock.  After more review I saw it's really referring to the middleware layer's MqttEvent processing callback.

I don't know how to better wordsmith that - maybe like `xMiddlewareCallback` - but if the cmocka layer could use that in function params and globals I think code would be easier to follow.

Reply:
Update the name to make it little better to understand the flow